tools/downloader/src/open_model_zoo/model_tools/_configuration.py

```diff
@@ -361,15 +362,42 @@ def deserialize(cls, model, name, subdirectory):
 
             return cls(name, subdirectory, files, postprocessing, mo_args, framework,
                 description, license_url, precisions, quantization_output_precisions,
-                task_type, conversion_to_onnx_args)
+                task_type, conversion_to_onnx_args, composite_model_name)
+
+def check_composite_model_dir(model_dir):
+    if list(model_dir.glob('*/*/model.yml')):
+        raise DeserializationError(
+            '{} directory should not contain any model.yml files in any subdirectories '
+            'that are not direct children of the composite model directory'.format(model_dir))
+
+    if list(model_dir.glob('model.yml')):
+        raise DeserializationError('{} directory should not contain any model.yml files'.format(model_dir))
+
+    model_name = model_dir.name
+    model_stages = list(model_dir.glob('*/model.yml'))
+    for model in model_stages:
+        if not model.parent.name.startswith(f'{model_name}-'):
+            raise DeserializationError('Names of composite model parts should start with composite model name')
 
 def load_models(args):
     models = []
     model_names = set()
 
+    composite_models = []
+
+    for composite_model_config in sorted(_common.MODEL_ROOT.glob('**/composite-model.yml')):
+        check_composite_model_dir(composite_model_config.parent)
+        composite_models.append(composite_model_config.parent.name)
+
+        if composite_models[-1] in composite_models[:-1]:
+            raise DeserializationError(
+                'Duplicate composite model name "{}"'.format(composite_models[-1]))
+
     for config_path in sorted(_common.MODEL_ROOT.glob('**/model.yml')):
         subdirectory = config_path.parent.relative_to(_common.MODEL_ROOT)
 
+        composite_model_name = subdirectory.parent.name if subdirectory.parent.name in composite_models else None
```

Comment:
This check is not quite correct. For example, consider these files:

```
a/composite-model.yml
b/a/b/model.yml
```

Here, the grandparent of `model.yml` is `a`, but the model `b` does not belong to the `a` composite model.

Reply:
Fixed